### ISSUE TITLE
Sites: Attempt to reload sites list if un-listed site is requested

### ISSUE THREAD

**Issue by beaulebens** [Automattic/wp-calypso#5571]
Many Calypso URLs contain a site identifier, which loads content specific to a site which is related to the current (logged in) user. When a user adds a new site to their account (presumably through either a signup flow, or via Jetpack), we need to refresh their site list to ensure that these routes work, and that they are available to load content specific to the site in question.

During signup flows, which now occur in Calypso, this isn't a problem, because the site list can just be force-refreshed, so as soon as users access the URLs, they see their site details. For Jetpack, currently this _is_ a problem. A user will add a site, and then when they first visit Calypso, they will usually hit one of 2 errors:

1. They try to access a site-specific URL that doesn't support on-the-fly site selection, so they are redirected to `/stats/day` (stats summary for all sites), or
2. They hit a route that _does_ support on-the-fly site selection, so they are presented with a site picker which _doesn't_ contain the site they were expecting to see.

It seems like there are a few ways we could address this:

- Allow a parameter that could be passed in to force-refresh the site list before attempting to access the route. Seems open to abuse, creates a default bad experience, but would indeed solve the problem if used.
- When completing a Jetpack connection, attempt to trigger a refresh of the sites list pre-emptively? Not sure about this one. If it can work, it would solve the problem for Jetpack, but not for any other potential case.
- Change logic in Calypso a little, so that when someone tries to access a route for a site that's not in their list, the first thing Calypso does is attempt to refresh the site list, and do one more check against the new list. If it's still not there, then do whatever it would currently do (site picker or redirect), if it _is_ there now, then handle it the same as if it was always there.

Has anyone else come up against this? Any other possible solutions?

@dereksmart @richardmuscat @roccotripaldi @johnHackworth

**Comment by roccotripaldi** [comment]
> For Jetpack, currently this is a problem. A user will add a site, and then when they first visit Calypso, they will usually hit one of 2 errors

In this scenario, how did the user add a site?

**Comment by beaulebens** [comment]
@roccotripaldi: Just going through the normal Jetpack connection process. So they've installed Jetpack, and then clicked the "Connect" button in there, logged into/verified connection with a WP.com account, and landed back in their wp-admin/Jetpack. From there, a good example right now is if they click the button to "Compare Options" under "Premium Add-ons" (which points to a site-specific URL on wordpress.com). Usually, on a new connection, the site will not be found, so they'll be dropped back to a full-screen site picker instead.

**Comment by roccotripaldi** [comment]
Ah, got it. In that case, I think your third proposed solution may be the best. I look into that. This may become a non-issue if we decide to send all new connections through the sign up flow in Calypso. But it would be nice to have something in place in the mean time.

**Comment by johnHackworth** [comment]
The third solution is the best one by far. That being said, I'm not 100% sure how easy would be to implement, since sites list is in the middle of a redux-reimplementation and right now is a little... fragmented.

I think @aduth and @gwwar (and maybe @mtias ?) were the last ones who were working with the new sites' redux subtree... so ping for opinions :)

**Comment by mtias** [comment]
Sites still has a long way to finish being ported to redux and all controllers still depend on sites-list. Shouldn't be hard to do this in `siteSelection` function of my-sites controller.

Check the code here: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/controller.js#L176-L189

**Comment by mtias** [comment]
@johnHackworth you can also just use `controller.awaitSiteLoaded` in the specific routes you need, that's probably better.

**Comment by johnHackworth** [comment]
@mtias hmm, the scope here is "the specific routes you need" is potentially any route that works with a selected site. I guess we could narrow it down to a few selected routes, but the problem is pretty much general for any jetpack site (correct me if I'm wrong here, @beaulebens )

**Comment by beaulebens** [comment]
@johnHackworth yep, you're right, it could be "pretty much any route", although I believe right now, the list of routes would be relatively short. If it's easier to just tackle it by tweaking specific routes, I can get together a list of them, otherwise a generic/core fix would handle it for everything, right?

**Comment by mtias** [comment]
A generic "fix" would just mean we bypass the cache all the time, which doesn't seem great. Is the `site_count` data in user bootstrap up to date by that time? Perhaps you could check if sitesList sites match the number site_count gives, and if it doesn't, force app to wait.

**Comment by beaulebens** [comment]
I'm definitely not suggesting bypassing the cache _all_ the time. I'm saying to "bypass" if the site in the URL was _not_ in the cache. If it's there, it'd use the cached information.

**Comment by mtias** [comment]
Oh, yes, I just meant all the time if there wasn't a site matched, which would be removing line 178. I'm not sure if that would cause other issues, though, specially if we can rely on site_count to make the call.

**Comment by roccotripaldi** [comment]
I've started digging into this a little today. One thing I've noticed is that the errors described only happen when I'm testing with my a11n wordpress account. It does not happen with my non-a11n accounts. This leads me to believe that it is only affecting folks with large site lists. I still think it is a priority to fix it, but my guess is that it doesn't affect a lot of folks.

**Comment by richardmuscat** [comment]
@roccotripaldi: I've definitely had problems with WP.com links and IIRC also without a8c account. 

How do you define 'large number of sites'? And does that number refer only to Jetpack sites or Jetpack and WP.com sites combined?

One thing we've noticed is that only about 50% of clicks on the links to plan pages result in a page view of plans pages which leads me to think this is not an a11n only problem.

**Comment by mtias** [comment]
@roccotripaldi have you looked at whether `user.site_count` has the correct number? Because if that is the case we should do something at the framework level like @beaulebens suggests, but we can tailor it for the event when we have a mismatch:

```
if ( user.site_count !== sitesList.get().length ) {
    // Wait for a re-fetch of sites
}
```

Does that make sense?

**Comment by roccotripaldi** [comment]
@mtias - that does make sense. I'll let you know if `user.site_count` is correct

**Comment by roccotripaldi** [comment]
Comparing user.visible_site_count to sites.getVisible() seems more accurate. however, re-fetching does not seem to remedy. sometimes it takes one or two refetches for the newly connected site to show up. I'll keep digging. Maybe there is something that can be improved server side.

**Comment by roccotripaldi** [comment]
closed in #7250
